--- a/lib/cli/CMakeLists.txt
+++ b/lib/cli/CMakeLists.txt
@@ -45,7 +45,8 @@
 if(CLI_COMPILE_SHARED_LIB)
     add_library(cli-shared-lib SHARED $<TARGET_OBJECTS:cli-obj>)
     set_target_properties(cli-shared-lib PROPERTIES
-                                         OUTPUT_NAME cli
+                                         OUTPUT_NAME aff3ct-cli
+                                         SOVERSION 0
                                          POSITION_INDEPENDENT_CODE ON) # set -fpic
     if(NOT CLI_STATUS_MESSAGE_QUIET)
         message(STATUS "cli - Compile: shared library")
@@ -179,4 +180,4 @@
             DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/cli
             COMPONENT headers
             FILES_MATCHING PATTERN "*.hxx")
-endif()
\ No newline at end of file
+endif()
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -218,6 +218,17 @@
     check_pie_supported()
 endif()
 
+# Append {shared,static}-library or target objects
+function(append_subproject TARGET SUBPROJECT SUBPROJECT_VAR ALLOW_SHARED_LIB)
+    if(${${SUBPROJECT_VAR}_COMPILE_SHARED_LIB} AND ${ALLOW_SHARED_LIB})
+        target_link_libraries(${TARGET} PRIVATE ${SUBPROJECT}-shared-lib)
+    elseif(CLI_COMPILE_STATIC_LIB)
+        target_link_libraries(${TARGET} PRIVATE ${SUBPROJECT}-static-lib)
+    else()
+        target_sources(${TARGET} PRIVATE $<TARGET_OBJECTS:${SUBPROJECT}-obj>)
+    endif()
+endfunction()
+
 # Object
 if(AFF3CT_COMPILE_OBJ)
     add_library(aff3ct-bin-obj OBJECT ${bin_source_files})
@@ -232,10 +243,11 @@
 # Binary
 if(AFF3CT_COMPILE_EXE)
     if(AFF3CT_INCLUDE_SPU_LIB)
-        add_executable(aff3ct-bin $<TARGET_OBJECTS:aff3ct-bin-obj> $<TARGET_OBJECTS:cli-obj> $<TARGET_OBJECTS:spu-obj>)
+        add_executable(aff3ct-bin $<TARGET_OBJECTS:aff3ct-bin-obj> $<TARGET_OBJECTS:spu-obj>)
     else()
-        add_executable(aff3ct-bin $<TARGET_OBJECTS:aff3ct-bin-obj> $<TARGET_OBJECTS:cli-obj>)
+        add_executable(aff3ct-bin $<TARGET_OBJECTS:aff3ct-bin-obj>)
     endif()
+    append_subproject(aff3ct-bin cli CLI ON)
     set_target_properties(aff3ct-bin PROPERTIES
                                      OUTPUT_NAME aff3ct${AFF3CT_EXE_SUFFIX}-${AFF3CT_VERSION_FULL}
                                      POSITION_INDEPENDENT_CODE ON) # set -fpie
@@ -245,10 +257,11 @@
 # Library
 if(AFF3CT_COMPILE_SHARED_LIB)
     if(AFF3CT_INCLUDE_SPU_LIB)
-        add_library(aff3ct-shared-lib SHARED $<TARGET_OBJECTS:aff3ct-lib-obj> $<TARGET_OBJECTS:cli-obj> $<TARGET_OBJECTS:spu-obj>)
+        add_library(aff3ct-shared-lib SHARED $<TARGET_OBJECTS:aff3ct-lib-obj> $<TARGET_OBJECTS:spu-obj>)
     else()
-        add_library(aff3ct-shared-lib SHARED $<TARGET_OBJECTS:aff3ct-lib-obj> $<TARGET_OBJECTS:cli-obj>)
+        add_library(aff3ct-shared-lib SHARED $<TARGET_OBJECTS:aff3ct-lib-obj>)
     endif()
+    append_subproject(aff3ct-shared-lib cli CLI ON)
     set_target_properties(aff3ct-shared-lib PROPERTIES
                                             OUTPUT_NAME aff3ct${AFF3CT_LIB_SUFFIX}-${AFF3CT_VERSION_FULL}
                                             POSITION_INDEPENDENT_CODE ON) # set -fpic
@@ -259,10 +272,11 @@
 endif(AFF3CT_COMPILE_SHARED_LIB)
 if(AFF3CT_COMPILE_STATIC_LIB)
     if(AFF3CT_INCLUDE_SPU_LIB)
-        add_library(aff3ct-static-lib STATIC $<TARGET_OBJECTS:aff3ct-lib-obj> $<TARGET_OBJECTS:cli-obj> $<TARGET_OBJECTS:spu-obj>)
+        add_library(aff3ct-static-lib STATIC $<TARGET_OBJECTS:aff3ct-lib-obj> $<TARGET_OBJECTS:spu-obj>)
     else()
-        add_library(aff3ct-static-lib STATIC $<TARGET_OBJECTS:aff3ct-lib-obj> $<TARGET_OBJECTS:cli-obj>)
+        add_library(aff3ct-static-lib STATIC $<TARGET_OBJECTS:aff3ct-lib-obj>)
     endif()
+    append_subproject(aff3ct-static-lib cli CLI OFF)
     set_target_properties(aff3ct-static-lib PROPERTIES
                                             OUTPUT_NAME aff3ct${AFF3CT_LIB_SUFFIX}-${AFF3CT_VERSION_FULL}
                                             POSITION_INDEPENDENT_CODE ON) # set -fpic
@@ -604,6 +618,17 @@
 # ------------------------------------------------------------------------------------------------------------- INSTALL
 # ---------------------------------------------------------------------------------------------------------------------
 
+if(CLI_COMPILE_SHARED_LIB)
+    install(TARGETS cli-shared-lib
+            LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}/
+            COMPONENT subproject-library)
+endif()
+if(CLI_COMPILE_STATIC_LIB)
+    install(TARGETS cli-static-lib
+            ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}/
+            COMPONENT subproject-library)
+endif()
+
 if(AFF3CT_COMPILE_EXE)
     install(TARGETS aff3ct-bin
             RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}/
