--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -136,7 +136,7 @@
 # Auto generate cmake config version file to link with AFF3CT library (only if an AFF3CT library has been compiled)
 if (AFF3CT_COMPILE_STATIC_LIB OR AFF3CT_COMPILE_SHARED_LIB)
     configure_file("${CMAKE_CURRENT_SOURCE_DIR}/cmake/Modules/aff3ct-config-version.cmake.in"
-                   "lib/cmake/aff3ct-${AFF3CT_VERSION_FULL}/aff3ct${AFF3CT_LIB_SUFFIX}-config-version.cmake" @ONLY)
+                   "lib/cmake/aff3ct${AFF3CT_LIB_SUFFIX}-${AFF3CT_VERSION_FULL}/aff3ct${AFF3CT_LIB_SUFFIX}-config-version.cmake" @ONLY)
 endif()
 
 # ------------------------------------------------------------------------------------------------------------- STRINGS
@@ -540,14 +540,14 @@
     export(TARGETS
            aff3ct-shared-lib
            NAMESPACE aff3ct${AFF3CT_LIB_SUFFIX}::
-           FILE "lib/cmake/aff3ct-${AFF3CT_VERSION_FULL}/aff3ct${AFF3CT_LIB_SUFFIX}-config.cmake")
+           FILE "lib/cmake/aff3ct${AFF3CT_LIB_SUFFIX}-${AFF3CT_VERSION_FULL}/aff3ct${AFF3CT_LIB_SUFFIX}-config.cmake")
 endif()
 
 if (AFF3CT_COMPILE_STATIC_LIB AND NOT AFF3CT_COMPILE_SHARED_LIB)
     export(TARGETS
            aff3ct-static-lib
            NAMESPACE aff3ct${AFF3CT_LIB_SUFFIX}::
-           FILE "lib/cmake/aff3ct-${AFF3CT_VERSION_FULL}/aff3ct${AFF3CT_LIB_SUFFIX}-config.cmake")
+           FILE "lib/cmake/aff3ct${AFF3CT_LIB_SUFFIX}$-${AFF3CT_VERSION_FULL}/aff3ct${AFF3CT_LIB_SUFFIX}-config.cmake")
 endif()
 
 if(AFF3CT_COMPILE_SHARED_LIB AND AFF3CT_COMPILE_STATIC_LIB)
@@ -555,7 +555,7 @@
            aff3ct-shared-lib
            aff3ct-static-lib
            NAMESPACE aff3ct${AFF3CT_LIB_SUFFIX}::
-           FILE "lib/cmake/aff3ct-${AFF3CT_VERSION_FULL}/aff3ct${AFF3CT_LIB_SUFFIX}-config.cmake")
+           FILE "lib/cmake/aff3ct${AFF3CT_LIB_SUFFIX}-${AFF3CT_VERSION_FULL}/aff3ct${AFF3CT_LIB_SUFFIX}-config.cmake")
 endif()
 
 # ---------------------------------------------------------------------------------------------------------------------
@@ -619,12 +619,12 @@
 if (AFF3CT_COMPILE_SHARED_LIB OR AFF3CT_COMPILE_STATIC_LIB)
     install(EXPORT
             aff3ct${AFF3CT_LIB_SUFFIX}-config
-            DESTINATION "${CMAKE_INSTALL_LIBDIR}/cmake/aff3ct-${AFF3CT_VERSION_FULL}/"
+            DESTINATION "${CMAKE_INSTALL_LIBDIR}/cmake/aff3ct${AFF3CT_LIB_SUFFIX}-${AFF3CT_VERSION_FULL}/"
             NAMESPACE aff3ct${AFF3CT_LIB_SUFFIX}::
             COMPONENT library)
 
-    install(FILES "${CMAKE_CURRENT_BINARY_DIR}/lib/cmake/aff3ct-${AFF3CT_VERSION_FULL}/aff3ct${AFF3CT_LIB_SUFFIX}-config-version.cmake"
-            DESTINATION "${CMAKE_INSTALL_LIBDIR}/cmake/aff3ct-${AFF3CT_VERSION_FULL}/"
+    install(FILES "${CMAKE_CURRENT_BINARY_DIR}/lib/cmake/aff3ct${AFF3CT_LIB_SUFFIX}-${AFF3CT_VERSION_FULL}/aff3ct${AFF3CT_LIB_SUFFIX}-config-version.cmake"
+            DESTINATION "${CMAKE_INSTALL_LIBDIR}/cmake/aff3ct${AFF3CT_LIB_SUFFIX}-${AFF3CT_VERSION_FULL}/"
             COMPONENT library)
 
     install(DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/include/"
